sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java

```diff
@@ -993,57 +994,86 @@ private <T> Flux<FeedResponse<T>> createQueryInternal(
         }, Queues.SMALL_BUFFER_SIZE, 1);
     }
 
+    private static void applyExceptionToMergedDiagnostics(
+        CosmosQueryRequestOptions requestOptions,
+        CosmosException exception) {
+
+        List<CosmosDiagnostics> cancelledRequestDiagnostics =
+            ImplementationBridgeHelpers
+                .CosmosQueryRequestOptionsHelper
+                .getCosmosQueryRequestOptionsAccessor()
+                .getCancelledRequestDiagnosticsTracker(requestOptions);
+
+        // if there is any cancelled requests, collect cosmos diagnostics
+        if (cancelledRequestDiagnostics != null && !cancelledRequestDiagnostics.isEmpty()) {
+            // combine all the cosmos diagnostics
+            CosmosDiagnostics aggregratedCosmosDiagnostics =
+                cancelledRequestDiagnostics
+                    .stream()
+                    .reduce((first, toBeMerged) -> {
+                        ClientSideRequestStatistics clientSideRequestStatistics =
+                            ImplementationBridgeHelpers
+                                .CosmosDiagnosticsHelper
+                                .getCosmosDiagnosticsAccessor()
+                                .getClientSideRequestStatisticsRaw(first);
+
+                        ClientSideRequestStatistics toBeMergedClientSideRequestStatistics =
+                            ImplementationBridgeHelpers
+                                .CosmosDiagnosticsHelper
+                                .getCosmosDiagnosticsAccessor()
+                                .getClientSideRequestStatisticsRaw(first);
+
+                        if (clientSideRequestStatistics == null) {
+                            return toBeMerged;
+                        } else {
+                            clientSideRequestStatistics.mergeClientSideRequestStatistics(toBeMergedClientSideRequestStatistics);
+                            return first;
+                        }
+                    })
+                    .get();
+
+            BridgeInternal.setCosmosDiagnostics(exception, aggregratedCosmosDiagnostics);
+        }
+    }
+
     private static <T> Flux<FeedResponse<T>> getFeedResponseFluxWithTimeout(
         Flux<FeedResponse<T>> feedResponseFlux,
         CosmosEndToEndOperationLatencyPolicyConfig endToEndPolicyConfig,
         CosmosQueryRequestOptions requestOptions,
         final AtomicBoolean isQueryCancelledOnTimeout) {
 
+        Duration endToEndTimeout = endToEndPolicyConfig.getEndToEndOperationTimeout();
+
+
+        Flux<FeedResponse<T>> flux;
```

Comment:
No blocking and nitpicking - the logic of handling negative timeout seems pretty much the same, so maybe instead we add the negative timeout checking around line 1054.